pyroute2/wiset.py

```diff
@@ -0,0 +1,477 @@
+
+    def flush(self):
+        """ Flush entries of the ipset """
+        flush_ipset(self.name, socket=self.socket)
+
+    @property
+    def content(self):
+        """ Dictionary of entries in the set.
+
+        Keys are IP addresses (as string), values are IPStats tuples.
+        """
+        if self._content is None:
+            self.update_content()
+
+        return self._content
+
+    def insert_list(self, entries):
+        """ Just a small helper to reduce the number of loops in main code. """
+        for entry in entries:
+            self.add(entry)
+
+    def replace_entries(self, new_list):
+        """ Replace the content of an ipset with a new list of entries.
+
+        This operation is like a flush() and adding all entries one by one. But
+        this call is atomic: it creates a temporary ipset and swap the content.
+
+        :param new_list: list of entries to add
+        :type new_list: list or :py:class:`set`
+        """
+        temp_name = str(uuid.uuid4())[0:8]
+        # Get a copy of ourself
+        temp = list_ipset(self.name, socket=self.socket)
+        temp.name = temp_name
+        temp.socket = self.socket
+        temp.create()
+        temp.insert_list(new_list)
+        swap_ipsets(self.name, temp_name, socket=self.socket)
+        temp.destroy()
+
+
+@need_ipset_socket
+def create_ipset(name, stype=None, family=AF_INET, exclusive=False,
+                 socket=None, **kwargs):
+    """ Create an ipset. """
+    socket.create(name, stype=stype, family=family, exclusive=exclusive,
+                  **kwargs)
+
+
+@need_ipset_socket
+def list_ipsets(content=False, socket=None, inherit_socket=False, prefix=None):
+    """ List all ipset as Wiset objects.
+
+    Get full ipset data from kernel and parse it in Wiset objects. Result is
+    a dictionary with ipset names as keys, and Wiset objects as values.
+
+    :param content: parse the list of entries and fill it in WiSet content
+                    dictionary
+    :type content: bool
+    :param inherit_socket: use the netlink socket passed in ipset arg to
+                           fill Wisets socket
+    :type inherit_socket: bool
+    :param prefix: filter out all ipset with a name not beginning by this prefix
+    :type prefix: str or None
+    """
+    res = {}
+    for myset in socket.list():
+        # on large sets, we can receive data in several messages
+        name = myset.get_attr("IPSET_ATTR_SETNAME")
+        if prefix is not None and not name.startswith(prefix):
+            continue
+        if name not in res:
+            wiset = WiSet.from_netlink(myset, content=content)
+            if inherit_socket:
+                wiset.socket = socket
+            res[wiset.name] = wiset
+        elif content:
+            res[wiset.name].update_dict_content(myset)
+    return res
+
+
+@need_ipset_socket
+def list_ipset(name, content=False, socket=None, inherit_socket=False):
+    """ Get one ipset as Wiset object
```

Comment:
A stupid comment — would it be possible to name it like `load_ipset()`? (just as a proposal to be discussed)

Another note: from my perspective it would be better not to name two different methods with only one letter difference in the names — too easy to make a mistake. I would propose `load_ipset()` and `load_all_ipsets()` or, at least, `list_ipset()` and `list_all_ipsets()`

Reply:
Yes, your are right. I will probably rename "socket" everywhere as well, since it can be confusing with the main "import socket" module.

Reply:
+1 about socket also.